jbpm-event-emitters/jbpm-event-emitters-elasticsearch/src/main/java/org/jbpm/event/emitters/elasticsearch/ElasticSearchEventEmitter.java

```diff
@@ -70,64 +64,69 @@
+            catch (ReflectiveOperationException | ClassCastException ex) {
+                logger.warn ("Error intializing configured factory "+className+ " using default",ex);
+            }
+        }
+        return factory != null ? factory : new DefaultESInstanceViewTransformerFactory();
+    }
+    
     private ObjectMapper mapper = new ObjectMapper();
+    
+    private ESInstanceViewTransformerFactory factory = getFactory();
+   
 
     private ExecutorService executor;
     
     private CloseableHttpClient httpclient;
 
     public ElasticSearchEventEmitter() {
-        mapper.setDateFormat(new SimpleDateFormat(dateFormatStr));
-        mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
-        mapper.configure(MapperFeature.PROPAGATE_TRANSIENT_MARKER, true);
-        
+        factory.configureObjectMapper(mapper);
         executor = buildExecutorService();
         httpclient = buildClient();
     }
+    
 
+    @Override
     public void deliver(Collection<InstanceView<?>> data) {
         // no-op
 
     }
 
+    @Override
     public void apply(Collection<InstanceView<?>> data) {
         if (data.isEmpty()) {
             return;
         }
 
         executor.execute(() -> {
             StringBuilder content = new StringBuilder();
-
+            Set<String> visitedIds = new HashSet<>();
             for (InstanceView<?> view : data) {
                 try {
-                    String json = mapper.writeValueAsString(view);
-
-                    String index = "jbpm";
-                    String type = "unknown";
-                    String id = "";
-                    if (view instanceof ProcessInstanceView) {
-                        index = "processes";
-                        type = "process";
-                        id = ((ProcessInstanceView) view).getCompositeId();
-                    } else if (view instanceof TaskInstanceView) {
-                        index = "tasks";
-                        type = "task";
-                        id = ((TaskInstanceView) view).getCompositeId();
-                    } else if (view instanceof CaseInstanceView) {
-                        index = "cases";
-                        type = "case";
-                        id = ((CaseInstanceView) view).getCompositeId();
+                    ESInstanceViewTransformer instanceView = factory.getInstanceViewTransfomer(view);
+                    if (instanceView == null) {
+                        logger.warn("Unsupported view {} ", view.getClass());
+                        continue;
                     }
-
-                    content.append("{ \"index\" : { \"_index\" : \"" + index + "\", \"_type\" : \"" + type + "\", \"_id\" : \"" + id + "\" } }\n");
-                    content.append(json);
-                    content.append("\n");
-
+                    ESRequest request = visitedIds.add(view.getCompositeId()) ? instanceView.index(view) : instanceView
+                            .update(view);
```

Comment:
Since `view` already contains all the event data, perhaps it's not necessary to differentiate between `index` and `update` operations. According to [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-request-body), `index` updates the document if it already exists.

Wdyt?

Reply:
The idea here is to merge the two views (task and operation) into the same index. 
The operation view does not really contain the same data than the task view, since eventTask source field is in fact transient. OperationView just contain type, userId and assingType (which will be null for all operations different that UPDATE list of users)

Reply:
OK I understand, thanks. Then I think the best approach would be sending `TaskInstanceView`s with the `index` operation and `TaskOperationView`s with `update`, right? Otherwise I can imagine that if a `TaskInstanceView` is sent with the `update` operation, it can leave fields out of sync (e.g. if some fields are removed)

Reply:
yes, thats what should be happening, index being used for task view and
update being used for tak operation view

Reply:
Is it safe to assume that the task view will always come first and then the task operation view? Even if it's 100% sure, perhaps we could make it more explicit (e.g. `if(view instance of TaskOperationView) {...`). This way the code would be easier to understand.

Reply:
Yes it is safe to assume that. 
This code should not perform instanceof, since it will stop being generic

Reply:
I understand. Nevertheless during local testing I've found some unexpected behavior. For example, I implemented a test emitter with this method:

```java
@Override
public void deliver(Collection<InstanceView<?>> data) {
    logger.debug("Preparing bulk delivery of {} instance views: {}", data.size(),
            data.stream().map(v -> v.getClass().getSimpleName()).collect(Collectors.joining(", ")));
    for (InstanceView<?> view : data) {
        if (!(view instanceof TaskInstanceView) || !(view instanceof ProcessInstanceView)) {
            logger.warn("Unsupported view class {}: {}", view.getClass().getSimpleName(), mapper.writeValueAsString(view));
        }
    }
}
```

I deployed the "evaluation" process that comes with the examples and started a process instance. Surprisingly this is the log the emitter produced:

```
Preparing bulk delivery of 4 instance views: TaskOperationView, ProcessInstanceView, TaskInstanceView, TaskOperationView
Unsupported view class TaskOperationView: {"compositeId":"jbpm-development-server_15","type":"ACTIVATED","assignType":null,"date":"2021-07-28T18:36:07.516+0000","userId":"jbpmAdmin"}
Unsupported view class TaskOperationView: {"compositeId":"jbpm-development-server_15","type":"ADDED","assignType":null,"date":"2021-07-28T18:36:07.510+0000","userId":"jbpmAdmin"}
```

In this case the first view received in the collection is a `TaskOperationView` instead of a `TaskInstanceView` (in case order mattered) and both `TaskOperationView` instances generated have a `compositeId`. Is this expected?